### PR TITLE
PT-3084: create 'carrier' gene status option (Genotype info>list of genes)

### PR DESCRIPTION
**Pull request by veronikaslc**

### CODE REVIEW

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java [hunk 1]
```diff
@@ -116,7 +118,7 @@
     private static final String JSON_REJECTEDGENES_KEY = "rejectedGenes";
 
     private static final List<String> STATUS_VALUES = Arrays.asList(INTERNAL_CANDIDATE_VALUE, INTERNAL_REJECTED_VALUE,
-        INTERNAL_SOLVED_VALUE);
+        INTERNAL_SOLVED_VALUE, INTERNAL_CARRIER_VALUE);
```
Comment (marta-) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java]: This comment has nothing to do with the 'carrier' issue, but hardcoding property values in the controller is wrong and, as we can see here, difficult to maintain. The value should be loaded from the XClass.

File: components/patient-data/indexing/pom.xml [hunk 2]
```diff
@@ -29,7 +29,7 @@
   <name>PhenoTips - Patient data - Solr indexing</name>
 
   <properties>
-    <coverage.instructionRatio>0.96</coverage.instructionRatio>
+    <coverage.instructionRatio>0.94</coverage.instructionRatio>
```
Comment (marta-) [inline components/patient-data/indexing/pom.xml]: :(

**Comment by marta-** [comment]
@veronikaslc I haven't tested yet, but as far as i understand from the code it doesn't look like carriers are represented in the pedigree according to the standard. They should be dots in the center of the individual's shape (or near the center, if an individual is a carrier for more than one gene), and the color of the dot for Gene X should always be the same as the color associated to confirmed causative Gene X (if this confirmed gene exists in the pedigree).

**Comment by veronikaslc** [comment]
@marta- didn't know that is needed, will do. We probably have to specify expected behaviour in more details.

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java (hunk 1 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java]: That's why I have a PR for GeneClass :)

Comment (marta-) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java]: But [the values are still hardcoded in that PR](https://github.com/phenotips/phenotips/pull/2216/files#diff-665baeaa42f5fff78d584ea090d111e2R61)...

**Comment by allasm** [comment]
@marta - I think we can implement pedigree graphic for gene carriers in v2. Currently the carrier "dot" graphic is overloaded, as it may refer to either genes, disorders or cancers. And since all of those will be modified in PhenoTips (cancers added, carrier status added for disorders, as far as I know?) it probably makes more sense to modify pedigree once all those are done.

Also if you mean to add dots in the middle of gene section (instead of one dot in the middle) that should probably be a separate task and done after the main "create 'carrier' gene status" task is competed

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java (hunk 1 above)
Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java]: It seems more logically to do it in in the PR for GeneClass.

File: components/patient-data/indexing/pom.xml (hunk 2 above)
Comment (veronikaslc) [inline components/patient-data/indexing/pom.xml]: added test, coverage restored

**Comment by marta-** [comment]
@veronikaslc @allasm okay

File: components/pedigree/resources/src/main/resources/pedigree/nodeMenuFields.js [hunk 3]
```diff
@@ -139,6 +139,13 @@ define([
                 'function' : 'setRejectedGenes' // not needed: read only for now
             },
             {
+                'name' : 'carrier_genes',
+                'label' : 'Genotype: carrier genes',
+                'type' : 'gene-picker',
+                'tab': 'Clinical',
+                'function' : 'setCarrierGenes' // not needed: read only for now
```
Comment (allasm) [inline components/pedigree/resources/src/main/resources/pedigree/nodeMenuFields.js]: wrong comment, it is not read-only?

File: components/pedigree/resources/src/main/resources/pedigree/view/carrierGeneLegend.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * @class CarrierGeneLegend
+ * @constructor
+ */
+ define(["pedigree/view/geneLegend"], function(GeneLegend){
+    var CarrierGeneLegend = Class.create( GeneLegend, {
+
+        initialize: function($super) {
+            $super('Carrier Genes', 'genes',
+                   "carrier",
+                   [], // these are never displayed in a legend so don't need colours
```
Comment (allasm) [inline components/pedigree/resources/src/main/resources/pedigree/view/carrierGeneLegend.js]: Are you sure we don't want to display them in a legend/on the node with colors?

File: components/pedigree/resources/src/main/resources/pedigree/view/person.js [hunk 5]
```diff
@@ -998,7 +1008,12 @@ define([
             return this._getGeneArray("rejected");
         },
 
-        // returns null if geneID is not presemnt, or gene status if it is
+        // used by controller in conjuntion with setCandidateGenes
```
Comment (allasm) [inline components/pedigree/resources/src/main/resources/pedigree/view/person.js]: wrong copy-paste comment

**Comment by allasm** [comment]
Based on the dev. meeting discussion we should make carrier genes editable in the pedigree (like candidate/solved but unlike negative which are read only)

File: components/pedigree/resources/src/main/resources/pedigree/nodeMenuFields.js (hunk 3 above)
Comment (veronikaslc) [inline components/pedigree/resources/src/main/resources/pedigree/nodeMenuFields.js]: it is read-only

File: components/pedigree/resources/src/main/resources/pedigree/view/carrierGeneLegend.js (hunk 4 above)
Comment (veronikaslc) [inline components/pedigree/resources/src/main/resources/pedigree/view/carrierGeneLegend.js]: in future

File: components/pedigree/resources/src/main/resources/pedigree/view/person.js (hunk 5 above)
Comment (veronikaslc) [inline components/pedigree/resources/src/main/resources/pedigree/view/person.js]: we have that comment for `getCausalGenes()` and `getRejectedGenes()`, if it is valid there why it is not valid for `getCarrierGenes()`?

**Comment by veronikaslc** [comment]
@marta- Did not we decide to make carrier genes editable it in future, not this PR?

**Comment by veronikaslc** [comment]
@allasm made a commit with codereview comments

**Comment by sashaandjic** [comment]
- carrier genes are not showing up in the legend and they are not coloured

**Comment by veronikaslc** [comment]
@sashaandjic comments addressed

**Comment by sashaandjic** [comment]
Tested! No problems found.

**Comment by sdumitriu** [comment]
@veronikaslc @marta- @allasm Is this ready to be re-reviewed?

**Comment by veronikaslc** [comment]
Yes it is ready.

**Pull request merged**